segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,179 @@
+     *
+     * @param writeLength The length of the write that completed.
+     */
+    void writeCompleted(int writeLength) {
+        synchronized (this) {
+            this.recentWriteCount++;
+            this.recentTotalWrittenLength += writeLength;
+        }
+    }
+
+    /**
+     * Gets a value indicating the current (most recent) degree of parallelism.
+     *
+     * @return The result
+     */
+    int getCurrentParallelism() {
+        return this.lastSnapshot.get().parallelism;
+    }
+
+    /**
+     * Updates the degree of parallelism (if needed), based on information collected since the last time this method was called.
+     * Rules:
+     * * Only changes something if throughput changed significantly since the last time.
+     * * Changes the degree of parallelism in the same direction as throughput.
+     * ** Exception is if both throughput and the fill ratio are declining, in which case nothing is done.
+     *
+     * @return The degree of parallelism, whether it was updated or not.
+     */
+    int updateParallelism() {
+        // Calculate the most recent throughput and Fill Ratio (of those items that have just been written).
+        final long time = this.timeSupplier.get();
+        final Snapshot lastSnapshot = this.lastSnapshot.get();
+        final long elapsedMillis = (time - lastSnapshot.timeStamp) / AbstractTimer.NANOS_TO_MILLIS;
+        if (elapsedMillis < MIN_FREQUENCY_MILLIS) {
+            // No need to do any change.
+            return lastSnapshot.parallelism;
+        } else if (elapsedMillis >= STALE_MILLIS) {
+            // Last snapshot is too old. We can't make any good determination.
+            resetSnapshot();
+            return this.lastSnapshot.get().parallelism;
+        }
+
+        final int recentCount;
+        final long recentLength;
+        synchronized (this) {
+            recentCount = this.recentWriteCount;
+            recentLength = this.recentTotalWrittenLength;
+            this.recentWriteCount = 0;
+            this.recentTotalWrittenLength = 0;
+        }
+
+        final double recentThroughput = (double) recentLength / elapsedMillis;
+        final double recentFillRatio = WriteQueue.calculateFillRatio(recentLength, recentCount);
+        final double throughputDifference = recentThroughput - lastSnapshot.throughput;
+
+        int parallelism = lastSnapshot.parallelism;
+        if (Math.abs(throughputDifference / lastSnapshot.throughput) >= SIGNIFICANT_DIFFERENCE) {
+            // Throughput changed significantly.
+            // But first, check for a special case. If both Throughput and Fill Ratio are decreasing, then most likely the
+            // former is a result of the latter. As such, don't do anything, in that case.
+            boolean fillRatioDecreased = (recentFillRatio - lastSnapshot.fillRatio) <= -SIGNIFICANT_DIFFERENCE;
+            if (throughputDifference > 0 || !fillRatioDecreased) {
```

Comment:
Similar to @fpj 's below I'm not sure the 'throughput' of an individual request is the right unit. It loses the distinction between a request that completed quickly and one that completed slowly but had a lot of data. A very important metric is the latency the write takes. In addition to this special case there are several others:
* If frames are unfilled and throughput has not changed we might want to increase or decrease the number depending on the latency.
* If frames are getting filled and throughput is decreasing because the calls are taking longer, we should do nothing.
* If throughput is decreasing because there is no data to send, then we shouldn't do anything, and should keep the old stats around lest traffic pick back up.
etc. 
Rather than dealing with them via a series of ifs, if we enumerated the constraints we need to satisfy I bet we could come up with a formula based on values tracked here.